src/test/java/com/google/devtools/build/lib/skyframe/TrackSourceDirectoriesIntegrationTest.java

```diff
@@ -0,0 +1,92 @@
+package com.google.devtools.build.lib.skyframe;
+
+import static com.google.common.truth.Truth.assertThat;
+import static com.google.devtools.build.lib.testutil.MoreAsserts.assertContainsEvent;
+import static com.google.devtools.build.lib.testutil.MoreAsserts.assertDoesNotContainEvent;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.devtools.build.lib.buildtool.util.BuildIntegrationTestCase;
+import com.google.devtools.build.lib.events.EventKind;
+import com.google.devtools.build.lib.events.util.EventCollectionApparatus;
+import com.google.devtools.build.lib.vfs.Path;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Tests for {@link TrackSourceDirectoriesFlag}. */
+@RunWith(JUnit4.class)
+public class TrackSourceDirectoriesIntegrationTest extends BuildIntegrationTestCase {
+
+  @Override
+  protected EventCollectionApparatus createEvents() {
+    return new EventCollectionApparatus(ImmutableSet.of(EventKind.FINISH));
+  }
+
+  @Before
+  public final void init() throws Exception {
+    // Enabled via jvm_flags.
+    assertThat(TrackSourceDirectoriesFlag.trackSourceDirectories()).isTrue();
+    addOptions("--experimental_ui_debug_all_events");
+
+    getWorkspace().getRelative("pkg/dir1/dir2").createDirectoryAndParents();
+    write("pkg/BUILD",
+        "genrule(",
+        "    name = 'a',",
+        "    srcs = ['dir1'],",
+        "    outs = ['out'],",
+        "    cmd = 'touch $@',",
+        ")");
+
+    // Build the target once - all tests below modify the source directory and assert that this
+    // causes the target to be rebuilt.
+    assertTargetRebuilds();
```

Comment:
I am not a huge fan of testing in `@Before`, but I can also see that as a sanity check. Can we rename this method to something more informative. Something like `setupAndBuildTestTarget`?